mdpow/analysis/dihedral.py

```diff
@@ -0,0 +1,67 @@
+# MDPOW: dihedral.py
+# 2021 Alia Lescoulie
+
+from typing import List
+
+import pandas as pd
+
+from MDAnalysis.exceptions import SelectionError
+from MDAnalysis.analysis.dihedrals import calc_dihedrals
+
+from .ensemble import EnsembleAtomGroup, EnsembleAnalysis
+
+import logging
+logger = logging.getLogger('mdpow.analysis.dihedral')
+
+
+class DihedralAnalysis(EnsembleAnalysis):
+    """Analyzes dihedral angles of selections from a single
+    :class:`~mdpow.analysis.ensemble.Ensemble` .
+
+    Accepts a list of :class:`~mdpow.analysis.ensemble.EnsembleAtomGroup`
+    with four atoms selected on each. All selections must be from the same
+    :class:`~mdpow.analysis.ensemble.Ensemble` .
+
+    Data is returned in a :class:`pandas.DataFrame` with observations sorted by
+    selection, solvent, interaction, lambda, time.
+    """
+
+    def __init__(self, dihedralgroups: List[EnsembleAtomGroup]):
+        super(DihedralAnalysis, self).__init__(dihedralgroups[0].ensemble)
+        self._sel = dihedralgroups
+        self._check_inputs()
+
+    def _check_inputs(self):
+        for i in range(len(self._sel) - 1):
+            # Checking if EnsembleAtomGroup.ensemble references same object in memory
+            if self._sel[i].ensemble is not self._sel[i - 1]:
+                logger.error('Dihedral selections from different Ensembles,'
+                             'ensure that all EnsembleAtomGroups are created'
+                             'from the same Ensemble.')
+                raise MemoryError
+        for group in self._sel:
+            for k in group.keys():
+                if len(group[k]) != 4:
+                    logger.error('Dihedral calculations require AtomGroups with'
+                                 'only 4 atoms, %s selected' % len(group))
+                    raise SelectionError
+
+    def _prepare_ensemble(self):
+        self._col = ['selection', 'solvent', 'interaction',
+                     'lambda', 'time', 'dihedral']
+        self.results = pd.DataFrame(columns=self._col)
+        self._res_dict = {key: [] for key in self._col}
+
+    def _single_frame(self):
+        for group in self._sel:
```

Comment:
I think I could reverse engineer that using `EnsembleAtomGroup.positions()` and reformatting the data into 4 `EnsembleAtomGroups`

Reply:
I have a bit of a complicated solution to this issue. I'll try to explain it here. I think this should work since it runs every dihedral in one universe simultaneously meaning that it very close to the same efficiency even with several dihedrals.

1. User give list of EnsembleAtomGroups
2. inputs are checked with `check_dihedral_inputs` and `check_group_from_ensemble`
3. `_reorg_groups` takes inputs and creates 4 `EnsembleAtomGroups` with each containing one the 4 atoms from each dihedral group. The system keys are `('selection', 'solvent', 'interaction', 'lambda').
4. In `_single_frame` the 4 new `EnsembleAtomGroup`'s keys for each selection are gotten, and used to get coordinates for the atoms at that key in the `Ensemble`. The coordinates are then used to calculate dihedrals.
5. The data is then placed in a dictionary with key structure from the dictionary
6.  Finally the data is organized into a `DataFrame`

Reply:
So far appears to work